src/main/java/org/takes/facets/hamcrest/HmRsHeader.java

```diff
@@ -24,55 +24,70 @@
 package org.takes.facets.hamcrest;
 
 import java.io.IOException;
+import java.util.ArrayList;
 import java.util.Collection;
 import java.util.Iterator;
-import java.util.LinkedList;
-import java.util.Locale;
-import java.util.Map;
-import java.util.Map.Entry;
+import java.util.List;
 import org.hamcrest.Description;
 import org.hamcrest.Matcher;
+import org.hamcrest.Matchers;
 import org.hamcrest.TypeSafeMatcher;
 import org.takes.Response;
-import org.takes.misc.EntryImpl;
 
 /**
  * Response Header Matcher.
  *
  * <p>This "matcher" tests given response header.
- * <p>The header name(s) should be provided in lowercase.
  * <p>The class is immutable and thread-safe.
  *
  * @author Eugene Kondrashev (eugene.kondrashev@gmail.com)
+ * @author I. Sokolov (happy.neko@gmail.com)
+ * @author Andrey Eliseev (aeg.exper0@gmail.com)
  * @version $Id$
  * @since 0.23.3
  */
 public final class HmRsHeader extends TypeSafeMatcher<Response> {
 
     /**
-     * Expected response header matcher.
+     * Values string used in description of mismatches.
      */
-    private final transient HeaderMatcher matcher;
+    private static final String VALUES_STR = " -> values: ";
+
+    /**
+     * Header matcher.
+     */
+    private final transient Matcher<String> headerm;
+
+    /**
+     * Value matcher.
+     */
+    private final transient Matcher<Iterable<String>> valuem;
+
+    /**
+     * Mismatched header values.
+     */
+    private transient Collection<String> failedvals;
 
     /**
      * Ctor.
-     * @param mtchr Matcher
+     * @param hdrm Header matcher
+     * @param vlm Value matcher
      */
-    public HmRsHeader(
-        final Matcher<? extends Map.Entry<String, String>> mtchr) {
```

Comment:
@exper0 we should probably keep this constructor for compatibility purposes and add a `@Deprecated` annotation to it.